### ISSUE TITLE
Rails server hangs up with concurrent json requests

### ISSUE THREAD

**Issue by sergiogomez** [rails/rails#32451]
### Steps to reproduce

First, create a new rails app with a simple scaffold:
```
rails new sample
cd sample
rails generate scaffold Post title:string
rails db:migrate
rails server
```
Then, launch two concurrent requests from another shell:
```
wget -O /dev/null \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  http://localhost:3000/posts & \
wget -O /dev/null \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  http://localhost:3000/posts
```
### Expected behavior
```
=> Booting Puma
=> Rails 5.1.6 application starting in development
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.11.3 (ruby 2.5.0-p0), codename: Love Song
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Use Ctrl-C to stop
Started GET "/posts" for 127.0.0.1 at 2018-04-04 16:03:06 +0200
Started GET "/posts" for 127.0.0.1 at 2018-04-04 16:03:06 +0200
   (0.4ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
   (0.2ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
Processing by PostsController#index as JSON
Processing by PostsController#index as JSON
  Rendering posts/index.json.jbuilder
  Rendering posts/index.json.jbuilder
  Post Load (0.4ms)  SELECT "posts".* FROM "posts"
  Post Load (0.1ms)  SELECT "posts".* FROM "posts"
  Rendered posts/index.json.jbuilder (4.0ms)
  Rendered posts/index.json.jbuilder (5.4ms)
Completed 200 OK in 24ms (Views: 18.3ms | ActiveRecord: 0.4ms)


Completed 200 OK in 32ms (Views: 19.1ms | ActiveRecord: 0.4ms)
```
### Actual behavior
Requests cause a hangup, and process must be killed (Ctrl+C is not working):

```
$ rails server
=> Booting Puma
=> Rails 5.1.6 application starting in development
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.11.3 (ruby 2.5.0-p0), codename: Love Song
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Use Ctrl-C to stop
Started GET "/posts" for 127.0.0.1 at 2018-04-04 13:06:40 +0200
Started GET "/posts" for 127.0.0.1 at 2018-04-04 13:06:40 +0200
   (0.3ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
   (0.1ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
```
If we add `ActionDispatch::DebugLocks` we have this info:

```
Thread 0 [0x3fd2fa1a9f48 sleep]  No lock (yielded share)
  Waiting in start_exclusive to "load"
  may be pre-empted for: "load"
  blocked by: 1

/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/monitor.rb:109:in `sleep'
/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/monitor.rb:109:in `wait'
/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/monitor.rb:109:in `wait'
/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/monitor.rb:121:in `wait_while'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/concurrency/share_lock.rb:219:in `wait_for'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/concurrency/share_lock.rb:81:in `block (2 levels) in start_exclusive'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/concurrency/share_lock.rb:185:in `yield_shares'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/concurrency/share_lock.rb:80:in `block in start_exclusive'
/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/concurrency/share_lock.rb:75:in `start_exclusive'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/concurrency/share_lock.rb:147:in `exclusive'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/dependencies/interlock.rb:11:in `loading'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/dependencies.rb:36:in `load_interlock'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/dependencies.rb:357:in `require_or_load'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/dependencies.rb:511:in `load_missing_constant'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/dependencies.rb:202:in `const_missing'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/inflector/methods.rb:269:in `const_get'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/inflector/methods.rb:269:in `block in constantize'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/inflector/methods.rb:267:in `each'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/inflector/methods.rb:267:in `inject'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/inflector/methods.rb:267:in `constantize'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/inflector/methods.rb:312:in `safe_constantize'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/core_ext/string/inflections.rb:77:in `safe_constantize'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:147:in `_default_wrap_model'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:94:in `block in model'
/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/mutex_m.rb:71:in `synchronize'
/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/mutex_m.rb:71:in `mu_synchronize'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:94:in `model'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:121:in `name'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:259:in `_wrapper_key'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:286:in `_wrapper_enabled?'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:235:in `process_action'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activerecord-5.1.6/lib/active_record/railties/controller_runtime.rb:22:in `process_action'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/abstract_controller/base.rb:124:in `process'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionview-5.1.6/lib/action_view/rendering.rb:30:in `process'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal.rb:189:in `dispatch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal.rb:253:in `dispatch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/routing/route_set.rb:49:in `dispatch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/routing/route_set.rb:31:in `serve'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/journey/router.rb:50:in `block in serve'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/journey/router.rb:33:in `each'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/journey/router.rb:33:in `serve'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/routing/route_set.rb:844:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/etag.rb:25:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/conditional_get.rb:25:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/head.rb:12:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/session/abstract/id.rb:232:in `context'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/session/abstract/id.rb:226:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/cookies.rb:613:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activerecord-5.1.6/lib/active_record/migration.rb:556:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/callbacks.rb:26:in `block in call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:97:in `run_callbacks'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/callbacks.rb:24:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/debug_exceptions.rb:59:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:135:in `call_app'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:28:in `block in call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:18:in `catch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:18:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/rack/logger.rb:36:in `call_app'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/rack/logger.rb:24:in `block in call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/tagged_logging.rb:69:in `block in tagged'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/tagged_logging.rb:26:in `tagged'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/tagged_logging.rb:69:in `tagged'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/rack/logger.rb:24:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/sprockets-rails-3.2.1/lib/sprockets/rails/quiet_assets.rb:13:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/request_id.rb:25:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/method_override.rb:22:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/runtime.rb:22:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/cache/strategy/local_cache_middleware.rb:27:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/static.rb:125:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/sendfile.rb:111:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/debug_locks.rb:39:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/engine.rb:522:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/configuration.rb:225:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/server.rb:624:in `handle_request'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/server.rb:438:in `process_client'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/server.rb:302:in `block in run'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/thread_pool.rb:120:in `block in spawn_thread'


---


Thread 1 [0x3fd2fa1a9ebc sleep]  Sharing
  blocking: 0

/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/mutex_m.rb:71:in `synchronize'
/Users/sergio/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/mutex_m.rb:71:in `mu_synchronize'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:94:in `model'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:121:in `name'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:259:in `_wrapper_key'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:286:in `_wrapper_enabled?'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal/params_wrapper.rb:235:in `process_action'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activerecord-5.1.6/lib/active_record/railties/controller_runtime.rb:22:in `process_action'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/abstract_controller/base.rb:124:in `process'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionview-5.1.6/lib/action_view/rendering.rb:30:in `process'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal.rb:189:in `dispatch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_controller/metal.rb:253:in `dispatch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/routing/route_set.rb:49:in `dispatch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/routing/route_set.rb:31:in `serve'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/journey/router.rb:50:in `block in serve'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/journey/router.rb:33:in `each'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/journey/router.rb:33:in `serve'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/routing/route_set.rb:844:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/etag.rb:25:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/conditional_get.rb:25:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/head.rb:12:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/session/abstract/id.rb:232:in `context'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/session/abstract/id.rb:226:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/cookies.rb:613:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activerecord-5.1.6/lib/active_record/migration.rb:556:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/callbacks.rb:26:in `block in call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:97:in `run_callbacks'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/callbacks.rb:24:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/debug_exceptions.rb:59:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:135:in `call_app'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:28:in `block in call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:18:in `catch'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/web-console-3.5.1/lib/web_console/middleware.rb:18:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/rack/logger.rb:36:in `call_app'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/rack/logger.rb:24:in `block in call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/tagged_logging.rb:69:in `block in tagged'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/tagged_logging.rb:26:in `tagged'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/tagged_logging.rb:69:in `tagged'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/rack/logger.rb:24:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/sprockets-rails-3.2.1/lib/sprockets/rails/quiet_assets.rb:13:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/request_id.rb:25:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/method_override.rb:22:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/runtime.rb:22:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/activesupport-5.1.6/lib/active_support/cache/strategy/local_cache_middleware.rb:27:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/static.rb:125:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/rack-2.0.4/lib/rack/sendfile.rb:111:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/actionpack-5.1.6/lib/action_dispatch/middleware/debug_locks.rb:39:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/railties-5.1.6/lib/rails/engine.rb:522:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/configuration.rb:225:in `call'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/server.rb:624:in `handle_request'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/server.rb:438:in `process_client'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/server.rb:302:in `block in run'
/Users/sergio/.rvm/gems/ruby-2.5.0@sample/gems/puma-3.11.3/lib/puma/thread_pool.rb:120:in `block in spawn_thread'
```

### System configuration
**Rails version**: 5.1.6

**Ruby version**: ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-darwin15]

**Comment by sergiogomez** [comment]
It seems related to autoload for development and `ActionController::ParamsWrapper`, because we can skip it disabling parameter wrapping for JSON from the initializer `config/initializers/wrap_parameters.rb`, and another workaround would be enabling `eager_load` code at development boot:

```
# config/environments/development.rb
config.eager_load = true
```

**Comment by sergiogomez** [comment]
With this change we could fix the issue, because it seems a problem between `synchronize` and `safe_constantize`:
https://github.com/rails/rails/compare/5-1-stable...sergiogomez:fix_concurrent_json_requests

**Comment by sergiogomez** [comment]
Hi @tenderlove @rafaelfranca , then, what do you think about the linked PRs (especially [the last one)](https://github.com/rails/rails/pull/32541), to solve this issue? Thanks!! 😉

**Comment by baurine** [comment]
it happens to me 100% when requesting multiple APIs at the same time from the client side, my environment is:

- Rails 5.2.1
- Ruby 2.5.1

the rails process hangs and `ctrl + c` doesn't work, you need to kill the puma process manually.

![wx20181120-092132](https://user-images.githubusercontent.com/1284531/48745502-972f8400-eca6-11e8-9281-b589919ddbc4.png)

![wx20181120-092312](https://user-images.githubusercontent.com/1284531/48745510-9d256500-eca6-11e8-8ce1-a9d091eae9c3.png)

I workarounded it by @sergiogomez 's suggestion, commented the `wrap_parameters format: [:json]` in `config/wrap_parameters.rb`, not sure whether it brings some side effects.

```
ActiveSupport.on_load(:action_controller) do
  # wrap_parameters format: [:json]
end
```

**Comment by nicklayb** [comment]
It's getting really annoying, every first concurrent request following a change on the backend makes it crash. I have to get to /home (where the notifications are fetched so a first unconcurrent requests occures) and then I can browse. I tried with `config.eager_load = true` it doesn't solve it. We dont have the `wrap_parameters` file either

**Comment by aurelienshz** [comment]
Another possible workaround if using puma  is to reduce the thread pool size to 1 in config/puma.rb:

```
threads_count = ENV.fetch("RAILS_MAX_THREADS") { 1 }
```

**Comment by fazo96** [comment]
We have the same problem. In development, rails keeps getting stuck seemingly at random. Using CTRL-C displays a graceful shutdown message but the process never ends unless force closed by CTRL-C.

Both workarounds provided here haven't changed anything

**Comment by reemanaik** [comment]
Hi, is this fix going to be back-ported to actionpack (5.2.3) ?
